### ISSUE TITLE
Alternative chaining syntax for Queue

### ISSUE THREAD

**Issue by tshemsedinov** [metarhia/metasync#44]
We have `ConcurrentQueue` in `lib/queue.js` but new chaining `metasync.queue` may implement more features from Queueing theory:
- Prioritized queueing
- Ordering: FIFO/LIFO
- Process throttling
- Balancing
- Timeouts
- Statistics and optimization
```
const cq = metasync.queue(3) // concurrency
  .wait(2000) // task may wait in queue not more then 2 sec
  .timeout(5000) // task processing timeout 5 sec
  .throttle(100, 1000) // throttle processing to 100 rps
  .process((item, cb) => cb(err, result)) // task processing function
  .success((item) => {}) // on success
  .failure((item) => {}) // on fail
  .drain(() => {}); // on drain
```
@aqrln @belochub @nechaido @DzyubSpirit maybe you need more in JSTP or can add something for GlobalStorage or Impress use cases?

**Comment by aqrln** [comment]
Right now I am working on other things, but such features should be useful when implementing a scheduler and load balancer. For the time being I cannot yet answer the question whether I will need more that these :)

**Comment by tshemsedinov** [comment]
@aqrln what I propose is to collect here ideas. For example:

We need to have three places to reject tasks:
- long waiting in queue
- long processing
- processing error

Tasks may have following fields:
- Source, destination and customer id/hash, to throttle certain customer, source, destination or their combinations (task routes). So processing can't be overloaded by one client or single target data piece accessed from multiple sources. I thing all those cases may be represented by single id/hash field.
- Task processing complexity (computational, memory, i/o)
- Task priority (for example 0-127 for applications, 128-255 for system)

**Comment by tshemsedinov** [comment]
This issue will be closed after merge PR #202 
But I extracted other features into: #206, #207 and #208
